### ISSUE TITLE
buggy behavior of BM upload during long CVR upload

### ISSUE THREAD

**Issue by sfsinger19103** [FreeAndFair/ColoradoRLA#788]
If the CVR file is uploaded first and while the circle is spinning it is possible to browse to the ballot manifest file, enter the hash value and click on the Upload button for the ballot manifest. The status for the ballot manifest file will falsely indicate that the ballot manifest file has been uploaded, but when the CVR file is finished being uploaded and the circle goes away the ballot manifest status reverts to not uploaded. The Upload and Re-upload buttons should be deactivated while an upload is being performed, especially for the very long upload times that the counties will see for the larger CVR files

from RLA Tool Feedback #1_tool version3.docx

**Comment by ranweiler** [comment]
I did some investigating, and was able to repro this consistently (using `tc` to simulate high upload latency/big file uploads).

1. Set latency.
1. Select both the ballot manifest and CVR export, enter the hashes, but do not click either upload button.
1. Click the CVR upload button.
1. Click the manifest upload button. 

The client *appears to be working as expected*. The reason it seems *not* to be is that, in the dashboard refresh JSON, I am getting both an uploaded ballot manifest (a non-empty `ballot_manifest_file` key) and an uploaded CVR export (a non-empty `cvr_export_file`), but the `asm_state` is `CVRS_OK`, which is wrong. Either the ASM state should be `BALLOT_MANIFEST_AND_CVRS_OK`, or the ballot manifest key should have no value. We're using the ASM state to decide if the ballot manifest has been uploaded (not the presence of some data), and so we observe an apparently empty form. We don't want to fix the display issue by checking for the data instead of the ASM, because then we'd be left in a state where it _looks_ (to the county user) like they uploaded all the necessary files, but to the DOS, it will look like they haven't imported a ballot manifest.

I suspect an ASM bug triggered by interleaved two-step uploads.

One trace of events that triggers the bug is as follows (as observed via client notifications):
1. CVR explort _uploaded_
1. Ballot manifest _uploaded_
1. Ballot manifest _imported_
1. CVR export _imported_

Relevant server logs:
```
21:02:20,854  INFO corla:399 - endpoint /upload-file hit by localhost:8888
21:02:20,854  INFO corla:147 - handling file upload request from 0:0:0:0:0:0:0:1
21:02:20,855  INFO corla:172 - successfully saved file of size 6187 from 0:0:0:0:0:0:0:1
21:02:21,805  INFO corla:399 - endpoint /contest/county hit by localhost:8888
21:02:21,866  INFO corla:399 - endpoint /upload-file hit by localhost:8888
21:02:21,866  INFO corla:147 - handling file upload request from 0:0:0:0:0:0:0:1
21:02:21,866  INFO corla:172 - successfully saved file of size 199 from 0:0:0:0:0:0:0:1
21:02:22,808  INFO corla:451 - successful /contest/county by countyadmin1 from localhost:8888
21:02:23,067  INFO corla:451 - successful /upload-file by countyadmin1 from localhost:8888
21:02:23,878  INFO corla:451 - successful /upload-file by countyadmin1 from localhost:8888
21:02:24,124  INFO corla:399 - endpoint /import-cvr-export hit by localhost:8888
21:02:24,177  INFO corla:399 - endpoint /contest/county hit by localhost:8888
21:02:24,729  INFO corla:399 - endpoint /import-ballot-manifest hit by localhost:8888
21:02:26,399  INFO corla:451 - successful /contest/county by countyadmin1 from localhost:8888
21:02:27,005  INFO corla:399 - endpoint /contest/county hit by localhost:8888
21:02:28,010  INFO corla:451 - successful /contest/county by countyadmin1 from localhost:8888
21:02:28,803  INFO corla:121 - 5 ballot manifest records parsed from file arapahoe-manifest.csv (id 40325)
21:02:29,006  INFO corla:399 - endpoint /contest/county hit by localhost:8888
21:02:29,206  INFO corla:242 - ASM event UPLOAD_BALLOT_MANIFEST_EVENT caused transition to BALLOT_MANIFEST_OK
21:02:29,809  INFO corla:451 - successful /import-ballot-manifest by countyadmin1 from localhost:8888
21:02:30,010  INFO corla:451 - successful /contest/county by countyadmin1 from localhost:8888
21:02:30,613  INFO corla:399 - endpoint /contest/county hit by localhost:8888
21:02:31,618  INFO corla:451 - successful /contest/county by countyadmin1 from localhost:8888
21:02:32,397  INFO corla:399 - endpoint /contest/county hit by localhost:8888
21:02:33,404  INFO corla:451 - successful /contest/county by countyadmin1 from localhost:8888
21:02:34,039  INFO corla:399 - endpoint /contest/county hit by localhost:8888
21:02:35,045  INFO corla:451 - successful /contest/county by countyadmin1 from localhost:8888
21:02:35,648  INFO corla:399 - endpoint /contest/county hit by localhost:8888
21:02:36,653  INFO corla:451 - successful /contest/county by countyadmin1 from localhost:8888
21:02:38,031  INFO corla:399 - endpoint /contest/county hit by localhost:8888
21:02:39,035  INFO corla:451 - successful /contest/county by countyadmin1 from localhost:8888
21:02:39,074  INFO corla:399 - endpoint /contest/county hit by localhost:8888
21:02:40,080  INFO corla:451 - successful /contest/county by countyadmin1 from localhost:8888
21:02:40,845  INFO corla:399 - endpoint /contest/county hit by localhost:8888
21:02:41,849  INFO corla:451 - successful /contest/county by countyadmin1 from localhost:8888
21:02:43,671  INFO corla:399 - endpoint /contest/county hit by localhost:8888
21:02:44,676  INFO corla:451 - successful /contest/county by countyadmin1 from localhost:8888
21:02:45,282  INFO corla:399 - endpoint /contest/county hit by localhost:8888
21:02:46,286  INFO corla:451 - successful /contest/county by countyadmin1 from localhost:8888
21:02:46,934  INFO corla:399 - endpoint /contest/county hit by localhost:8888
21:02:47,938  INFO corla:451 - successful /contest/county by countyadmin1 from localhost:8888
21:02:49,147  INFO corla:399 - endpoint /contest/county hit by localhost:8888
21:02:50,151  INFO corla:451 - successful /contest/county by countyadmin1 from localhost:8888
21:02:51,945  INFO corla:399 - endpoint /contest/county hit by localhost:8888
21:02:52,949  INFO corla:451 - successful /contest/county by countyadmin1 from localhost:8888
21:02:54,183  INFO corla:399 - endpoint /contest/county hit by localhost:8888
21:02:55,014  INFO corla:451 - successful /contest/county by countyadmin1 from localhost:8888
21:02:55,026  INFO corla:399 - endpoint /contest/county hit by localhost:8888
21:02:55,027  INFO corla:451 - successful /contest/county by countyadmin1 from localhost:8888
21:02:55,059  INFO corla:152 - 165 CVRs parsed from file 40324
21:02:55,063  INFO corla:242 - ASM event UPLOAD_CVRS_EVENT caused transition to CVRS_OK
21:02:55,063  INFO corla:451 - successful /import-cvr-export by countyadmin1 from localhost:8888
```

 cc @dmzimmerman

**Comment by dmzimmerman** [comment]
The root cause of this is the fact that CVR imports are multiple database transactions, and the ASM state is changing somewhere within those transactions. I _could_ try to fix this by instituting some sort of check that the ASM state read when ending an endpoint hit is the same as the ASM state read when the endpoint was started... but I'm not sure it would fix the problem in all cases. I think the easiest immediate solution is for the client to disallow interleaving these actions, but we can discuss it today (I'll investigate how long it'll take to implement the other fix; if it's super simple, I can do it quickly and you can test it).

**Comment by ranweiler** [comment]
Closed in #804.
